### PR TITLE
First stab at a go style guide

### PR DESCRIPTION
**Pull request by rickducott**

### CODE REVIEW

File: engineering/docs/go_conventions.md [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+type ChangelogReader interface {
+	GetChangelogForTag(ctx context.Context, tag string) (*Changelog, error)
+	ReadChangelogFile(ctx context.Context, path string) (*ChangelogFile, error)
+}
+
+type changelogReader struct {
+	code vfsutils.MountedRepo
+}
+
+func NewChangelogReader(code vfsutils.MountedRepo) ChangelogReader {
```
Comment (marcogschmidt) [inline engineering/docs/go_conventions.md]: I was recently made aware by @yuval-k of the "_accept interfaces, return structs_" best practice. There are different blogs which describe the advantages, but the basic idea is that it's the caller who should be able to define the interface it needs. I think it is a good rule of thumb and we should consider adopting it.

File: engineering/docs/go_conventions.md [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+Unit tests should be written for every new interface function implementation or public function. 
+Mockgen should be used as necessary to mock out underlying implementation dependencies. 
+Unit tests should be written for every error and interesting edge case. 
+
+### Tracking changes
+
+Repos should prefer to use changelogs and solobot for issue/PR linking and release management. 
+See https://github.com/solo-io/go-utils/tree/master/changelogutils.
+
+## Errors
```
Comment (marcogschmidt) [inline engineering/docs/go_conventions.md]: As a general note for the`error` section, I think we should start using the `New` and `Is` functions that wrap the correspondent ones in the new `xerror` package in the standard library. See [here](https://github.com/solo-io/go-utils/blob/1c2dd400703ec1a7340609a6f5c00653ecf3d5c7/errors/utils.go#L23)

File: engineering/docs/go_conventions.md (hunk 1 above)
Comment (rickducott) [reply engineering/docs/go_conventions.md]: I was under the (mistaken) impression that we couldn't return lower-cased structs and have them be accessible to other packages. I'm still not sure there's much of a difference to returning structs vs. interfaces in this context, but I'm happy to switch to that.

File: engineering/docs/go_conventions.md (hunk 2 above)
Comment (rickducott) [reply engineering/docs/go_conventions.md]: This is done in `github.com/solo-io/go-utils/errors`, which I'm happy to add here. I would like to see examples of "Is" being useful, however, because I think it is very underwhelming. 

```
expected := changelogutils.TooManyChangelogFilesAddedError(2)
file, err := validator.ValidateChangelog(ctx)
Expect(file).To(BeNil())
// This fails
Expect(errors.Is(err, expected)).To(BeTrue())
// This succeeds
Expect(err.Error()).To(Equal(expected.Error()))
```
